android-interop-testing/app/src/androidTest/java/io/grpc/android/integrationtest/TesterActivityTest.java

```diff
@@ -0,0 +1,58 @@
+package io.grpc.android.integrationtest;
+
+import android.os.Bundle;
+import android.test.ActivityInstrumentationTestCase2;
+import android.test.InstrumentationTestRunner;
+import junit.framework.Assert;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The Activity test of our App, used to trigger the integration test from commandline.
+ */
+public class TesterActivityTest extends ActivityInstrumentationTestCase2<TesterActivity> {
+
+  private String host;
+  private int port = 8030;
+  private String result;
+  private String serverHostOverride;
+  private String testCase;
+  private boolean useTls;
+  private boolean useTestCa;
+
+  public TesterActivityTest() {
+    super(TesterActivity.class);
+  }
+
+  public void setUp() throws Exception {
+    super.setUp();
+    getActivity();
+    Bundle args = ((InstrumentationTestRunner) getInstrumentation()).getArguments();
+    host = args.getString("server_host", "");
+    port = Integer.parseInt(args.getString("server_port", "0"));
+    assertTrue("Please specify server_host and server_port.", !host.isEmpty() && port != 0);
+    serverHostOverride = args.getString("server_host_override", null);
+    testCase = args.getString("test_case", "empty_unary");
+    useTls = Boolean.parseBoolean(args.getString("use_tls", "true"));
+    useTestCa = Boolean.parseBoolean(args.getString("use_test_ca", "false"));
+  }
+
+  public void testGrpc() throws Exception {
+    final CountDownLatch finished = new CountDownLatch(1);
+
+    new GrpcTestTask(testCase, host, port, serverHostOverride, useTls, useTestCa,
```

Comment:
Stepping back a bit.

For the interop testing, you don't need any _test_ classes. Instead, you can just use an activity and start it with [am start](http://developer.android.com/tools/help/adb.html#am). You can pass any necessary arguments as part of the intent (see the "Specification for \<INTENT\> arguments" section). Here's [an example](http://stackoverflow.com/a/3229077). This solves (1).

We could still have unit tests, just they would be run on Android. So you could use Http2OkHttpTest almost as-is on Android, and be able to test ALPN and other such things. To combine this with the previous paragraph, you could use a similar architecture to [TestServiceClient](https://github.com/grpc/grpc-java/blob/master/interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java).

For (2), you could use the constructor, or you could use a static block. But I don't actually see why continuing to use onCreate() wouldn't work. The `ProviderInstaller.installIfNeededAsync()` does seem like it should maybe be part of an [Application](http://developer.android.com/reference/android/app/Application.html) specified with <application android:name="classname"> in AndroidManifest.xml.